File: apps/mobile/native/android/com/j11/looper/mobile/SpokenFormatting.kt
package com.j11.looper.mobile

import java.util.Locale

internal object SpokenFormatting {
    private enum class Action {
        PUNCTUATION,
        NEWLINES,
        REMOVE_LAST_WORD,
        REMOVE_CLAUSE,
        OPEN,
        CLOSE,
        JOIN,
        BULLET,
        NUMBERED_ITEM,
        NEXT_ITEM,
        LITERAL_ON,
        LITERAL_OFF,
    }

    private enum class ListKind { BULLET, NUMBERED }

    private data class FormattingState(
        var literalMode: Boolean = false,
        var listKind: ListKind? = null,
        var listNumber: Int = 0,
        var joinNext: Boolean = false,
    )

    private data class Command(
        val languages: Set<String>,
        val phrase: List<String>,
        val action: Action,
        val value: String = "",
    )

    private val commands = listOf(
        Command(setOf("en"), listOf("end", "literal", "mode"), Action.LITERAL_OFF),
        Command(setOf("es"), listOf("fin", "modo", "literal"), Action.LITERAL_OFF),
        Command(setOf("pt"), listOf("fim", "modo", "literal"), Action.LITERAL_OFF),
        Command(setOf("en"), listOf("literal", "mode"), Action.LITERAL_ON),
        Command(setOf("es", "pt"), listOf("modo", "literal"), Action.LITERAL_ON),
        Command(setOf("en"), listOf("numbered", "item"), Action.NUMBERED_ITEM),
        Command(setOf("es"), listOf("elemento", "numerado"), Action.NUMBERED_ITEM),
        Command(setOf("pt"), listOf("item", "numerado"), Action.NUMBERED_ITEM),
        Command(setOf("en"), listOf("bullet", "point"), Action.BULLET),
        Command(setOf("es"), listOf("punto", "de", "lista"), Action.BULLET),
        Command(setOf("pt"), listOf("ponto", "de", "lista"), Action.BULLET),
        Command(setOf("en"), listOf("next", "item"), Action.NEXT_ITEM),
        Command(setOf("es"), listOf("siguiente", "elemento"), Action.NEXT_ITEM),
        Command(setOf("pt"), listOf("próximo", "item"), Action.NEXT_ITEM),
        Command(setOf("en"), listOf("open", "parenthesis"), Action.OPEN, "("),
        Command(setOf("es"), listOf("abre", "paréntesis"), Action.OPEN, "("),
        Command(setOf("pt"), listOf("abre", "parênteses"), Action.OPEN, "("),
        Command(setOf("en"), listOf("close", "parenthesis"), Action.CLOSE, ")"),
        Command(setOf("es"), listOf("cierra", "paréntesis"), Action.CLOSE, ")"),
        Command(setOf("pt"), listOf("fecha", "parênteses"), Action.CLOSE, ")"),
        Command(setOf("en"), listOf("open", "bracket"), Action.OPEN, "["),
        Command(setOf("es"), listOf("abre", "corchete"), Action.OPEN, "["),
        Command(setOf("pt"), listOf("abre", "colchete"), Action.OPEN, "["),
        Command(setOf("en"), listOf("close", "bracket"), Action.CLOSE, "]"),
        Command(setOf("es"), listOf("cierra", "corchete"), Action.CLOSE, "]"),
        Command(setOf("pt"), listOf("fecha", "colchete"), Action.CLOSE, "]"),
        Command(setOf("en"), listOf("at", "sign"), Action.JOIN, "@"),
        Command(setOf("es", "pt"), listOf("arroba"), Action.JOIN, "@"),
        Command(setOf("en"), listOf("dot"), Action.JOIN, "."),
        Command(setOf("es"), listOf("punto", "sin", "espacio"), Action.JOIN, "."),
        Command(setOf("pt"), listOf("ponto", "sem", "espaço"), Action.JOIN, "."),
        Command(setOf("en"), listOf("slash"), Action.JOIN, "/"),
        Command(setOf("es", "pt"), listOf("barra"), Action.JOIN, "/"),
        Command(setOf("es"), listOf("signo", "de", "interrogación"), Action.PUNCTUATION, "?"),
        Command(setOf("es"), listOf("signo", "de", "exclamación"), Action.PUNCTUATION, "!"),
        Command(setOf("pt"), listOf("ponto", "de", "interrogação"), Action.PUNCTUATION, "?"),
        Command(setOf("pt"), listOf("ponto", "de", "exclamação"), Action.PUNCTUATION, "!"),
        Command(setOf("en"), listOf("exclamation", "point"), Action.PUNCTUATION, "!"),
        Command(setOf("en"), listOf("exclamation", "mark"), Action.PUNCTUATION, "!"),
        Command(setOf("en"), listOf("question", "mark"), Action.PUNCTUATION, "?"),
        Command(setOf("en"), listOf("full", "stop"), Action.PUNCTUATION, "."),
        Command(setOf("en"), listOf("new", "paragraph"), Action.NEWLINES, "\n\n"),
        Command(setOf("es"), listOf("nuevo", "párrafo"), Action.NEWLINES, "\n\n"),
        Command(setOf("pt"), listOf("novo", "parágrafo"), Action.NEWLINES, "\n\n"),
        Command(setOf("en"), listOf("new", "line"), Action.NEWLINES, "\n"),
        Command(setOf("es"), listOf("nueva", "línea"), Action.NEWLINES, "\n"),
        Command(setOf("pt"), listOf("nova", "linha"), Action.NEWLINES, "\n"),
        Command(setOf("es"), listOf("punto", "y", "coma"), Action.PUNCTUATION, ";"),
        Command(setOf("pt"), listOf("ponto", "e", "vírgula"), Action.PUNCTUATION, ";"),
        Command(setOf("es"), listOf("dos", "puntos"), Action.PUNCTUATION, ":"),
        Command(setOf("pt"), listOf("dois", "pontos"), Action.PUNCTUATION, ":"),
        Command(setOf("en"), listOf("no", "wait"), Action.REMOVE_LAST_WORD),
        Command(setOf("en"), listOf("i", "mean"), Action.REMOVE_LAST_WORD),
        Command(setOf("es"), listOf("mejor", "dicho"), Action.REMOVE_LAST_WORD),
        Command(setOf("es"), listOf("quise", "decir"), Action.REMOVE_LAST_WORD),
        Command(setOf("pt"), listOf("quer", "dizer"), Action.REMOVE_LAST_WORD),
        Command(setOf("pt"), listOf("melhor", "dizendo"), Action.REMOVE_LAST_WORD),
        Command(setOf("en"), listOf("scratch", "that"), Action.REMOVE_CLAUSE),
        Command(setOf("es"), listOf("borra", "eso"), Action.REMOVE_CLAUSE),
        Command(setOf("pt"), listOf("apaga", "isso"), Action.REMOVE_CLAUSE),
        Command(setOf("en"), listOf("comma"), Action.PUNCTUATION, ","),
        Command(setOf("en"), listOf("period"), Action.PUNCTUATION, "."),
        Command(setOf("en"), listOf("colon"), Action.PUNCTUATION, ":"),
        Command(setOf("en"), listOf("semicolon"), Action.PUNCTUATION, ";"),
        Command(setOf("es"), listOf("coma"), Action.PUNCTUATION, ","),
        Command(setOf("es"), listOf("punto"), Action.PUNCTUATION, "."),
        Command(setOf("pt"), listOf("vírgula"), Action.PUNCTUATION, ","),
        Command(setOf("pt"), listOf("ponto"), Action.PUNCTUATION, "."),
    )

    fun apply(transcript: String, language: String): String {
        val rawTokens = transcript.trim().split(Regex("\\s+")).filter(String::isNotEmpty)
        if (rawTokens.isEmpty()) return ""
        val tokens = rawTokens.map(::normalizeToken)
        val languageCode = languageCode(language)
        val output = StringBuilder(transcript.length)
        val state = FormattingState()
        var index = 0

        while (index < rawTokens.size) {
            val command = commands.firstOrNull { candidate ->
                (languageCode == null || languageCode in candidate.languages) &&
                    index + candidate.phrase.size <= tokens.size &&
                    tokens.subList(index, index + candidate.phrase.size) == candidate.phrase
            }
            if (command != null) {
                if (state.literalMode) {
                    if (command.action == Action.LITERAL_OFF) {
                        state.literalMode = false
                        index += command.phrase.size
                        continue
                    }
                } else {
                    if (command.action == Action.LITERAL_ON) {
                        state.literalMode = true
                        index += command.phrase.size
                        continue
                    }
                    applyCommand(output, command, state)
                    index += command.phrase.size
                    continue
                }
            }

            if (!state.literalMode && languageCode != null) {
                val entity = SpokenEntities.match(tokens, index, languageCode)
                if (entity != null) {
                    if (output.isNotEmpty() && !output.last().isWhitespace() && !state.joinNext) {
                        output.append(' ')
                    }
                    output.append(entity.text)
                    state.joinNext = false
                    index += entity.consumed
                    continue
                }
            }

            if (output.isNotEmpty() && !output.last().isWhitespace() && !state.joinNext) {
                output.append(' ')
            }
            output.append(rawTokens[index])
            state.joinNext = false
            index += 1
        }
        return output.toString().trim()
    }

    private fun normalizeToken(token: String): String =
        token.trim { !it.isLetterOrDigit() }.lowercase(Locale.ROOT)

    private fun languageCode(language: String): String? {
        val normalized = language.trim().lowercase(Locale.ROOT)
        if (normalized.isEmpty() || normalized == "auto" || normalized.length < 2) return null
        return normalized.take(2).takeIf { it in setOf("en", "es", "pt") }
    }

    private fun trimEnd(output: StringBuilder) {
        while (output.isNotEmpty() && output.last().isWhitespace()) {
            output.deleteCharAt(output.lastIndex)
        }
    }

    private fun applyCommand(
        output: StringBuilder,
        command: Command,
        state: FormattingState,
    ) {
        when (command.action) {
            Action.PUNCTUATION -> {
                trimEnd(output)
                if (output.isNotEmpty() && !output.endsWith(command.value)) output.append(command.value)
            }
            Action.NEWLINES -> {
                trimEnd(output)
                if (output.isNotEmpty()) output.append(command.value)
            }
            Action.REMOVE_LAST_WORD -> {
                trimEnd(output)
                while (output.isNotEmpty() && !output.last().isWhitespace()) {
                    output.deleteCharAt(output.lastIndex)
                }
                trimEnd(output)
            }
            Action.REMOVE_CLAUSE -> {
                trimEnd(output)
                val keep = output.indexOfLast { it == '.' || it == '!' || it == '?' || it == '\n' }
                output.setLength(if (keep < 0) 0 else keep + 1)
                trimEnd(output)
            }
            Action.OPEN -> {
                trimEnd(output)
                if (output.isNotEmpty()) output.append(' ')
                output.append(command.value)
                state.joinNext = true
            }
            Action.CLOSE -> {
                trimEnd(output)
                output.append(command.value)
            }
            Action.JOIN -> {
                trimEnd(output)
                output.append(command.value)
                state.joinNext = true
            }
            Action.BULLET -> {
                state.listKind = ListKind.BULLET
                state.listNumber = 0
                startListItem(output, "- ")
            }
            Action.NUMBERED_ITEM -> {
                if (state.listKind != ListKind.NUMBERED) state.listNumber = 0
                state.listKind = ListKind.NUMBERED
                state.listNumber += 1
                startListItem(output, "${state.listNumber}. ")
            }
            Action.NEXT_ITEM -> {
                if (state.listKind == ListKind.NUMBERED) {
                    state.listNumber += 1
                    startListItem(output, "${state.listNumber}. ")
                } else {
                    state.listKind = ListKind.BULLET
                    startListItem(output, "- ")
                }
            }
            Action.LITERAL_ON, Action.LITERAL_OFF -> Unit
        }
    }

    private fun startListItem(output: StringBuilder, prefix: String) {
        trimEnd(output)
        if (output.isNotEmpty()) output.append('\n')
        output.append(prefix)
    }
}

File: apps/mobile/targets/keyboard/SpokenFormatting.swift
import Foundation

enum SpokenFormatting {
    private enum Action {
        case punctuation(Character)
        case newlines(String)
        case removeLastWord
        case removeClause
        case open(Character)
        case close(Character)
        case join(String)
        case bullet
        case numberedItem
        case nextItem
        case literalOn
        case literalOff
    }

    private struct Command {
        let languages: Set<String>
        let phrase: [String]
        let action: Action
    }

    private static let commands: [Command] = [
        Command(languages: ["en"], phrase: ["end", "literal", "mode"], action: .literalOff),
        Command(languages: ["es"], phrase: ["fin", "modo", "literal"], action: .literalOff),
        Command(languages: ["pt"], phrase: ["fim", "modo", "literal"], action: .literalOff),
        Command(languages: ["en"], phrase: ["literal", "mode"], action: .literalOn),
        Command(languages: ["es", "pt"], phrase: ["modo", "literal"], action: .literalOn),
        Command(languages: ["en"], phrase: ["numbered", "item"], action: .numberedItem),
        Command(languages: ["es"], phrase: ["elemento", "numerado"], action: .numberedItem),
        Command(languages: ["pt"], phrase: ["item", "numerado"], action: .numberedItem),
        Command(languages: ["en"], phrase: ["bullet", "point"], action: .bullet),
        Command(languages: ["es"], phrase: ["punto", "de", "lista"], action: .bullet),
        Command(languages: ["pt"], phrase: ["ponto", "de", "lista"], action: .bullet),
        Command(languages: ["en"], phrase: ["next", "item"], action: .nextItem),
        Command(languages: ["es"], phrase: ["siguiente", "elemento"], action: .nextItem),
        Command(languages: ["pt"], phrase: ["próximo", "item"], action: .nextItem),
        Command(languages: ["en"], phrase: ["open", "parenthesis"], action: .open("(")),
        Command(languages: ["es"], phrase: ["abre", "paréntesis"], action: .open("(")),
        Command(languages: ["pt"], phrase: ["abre", "parênteses"], action: .open("(")),
        Command(languages: ["en"], phrase: ["close", "parenthesis"], action: .close(")")),
        Command(languages: ["es"], phrase: ["cierra", "paréntesis"], action: .close(")")),
        Command(languages: ["pt"], phrase: ["fecha", "parênteses"], action: .close(")")),
        Command(languages: ["en"], phrase: ["open", "bracket"], action: .open("[")),
        Command(languages: ["es"], phrase: ["abre", "corchete"], action: .open("[")),
        Command(languages: ["pt"], phrase: ["abre", "colchete"], action: .open("[")),
        Command(languages: ["en"], phrase: ["close", "bracket"], action: .close("]")),
        Command(languages: ["es"], phrase: ["cierra", "corchete"], action: .close("]")),
        Command(languages: ["pt"], phrase: ["fecha", "colchete"], action: .close("]")),
        Command(languages: ["en"], phrase: ["at", "sign"], action: .join("@")),
        Command(languages: ["es", "pt"], phrase: ["arroba"], action: .join("@")),
        Command(languages: ["en"], phrase: ["dot"], action: .join(".")),
        Command(languages: ["es"], phrase: ["punto", "sin", "espacio"], action: .join(".")),
        Command(languages: ["pt"], phrase: ["ponto", "sem", "espaço"], action: .join(".")),
        Command(languages: ["en"], phrase: ["slash"], action: .join("/")),
        Command(languages: ["es", "pt"], phrase: ["barra"], action: .join("/")),
        Command(languages: ["es"], phrase: ["signo", "de", "interrogación"], action: .punctuation("?")),
        Command(languages: ["es"], phrase: ["signo", "de", "exclamación"], action: .punctuation("!")),
        Command(languages: ["pt"], phrase: ["ponto", "de", "interrogação"], action: .punctuation("?")),
        Command(languages: ["pt"], phrase: ["ponto", "de", "exclamação"], action: .punctuation("!")),
        Command(languages: ["en"], phrase: ["exclamation", "point"], action: .punctuation("!")),
        Command(languages: ["en"], phrase: ["exclamation", "mark"], action: .punctuation("!")),
        Command(languages: ["en"], phrase: ["question", "mark"], action: .punctuation("?")),
        Command(languages: ["en"], phrase: ["full", "stop"], action: .punctuation(".")),
        Command(languages: ["en"], phrase: ["new", "paragraph"], action: .newlines("\n\n")),
        Command(languages: ["es"], phrase: ["nuevo", "párrafo"], action: .newlines("\n\n")),
        Command(languages: ["pt"], phrase: ["novo", "parágrafo"], action: .newlines("\n\n")),
        Command(languages: ["en"], phrase: ["new", "line"], action: .newlines("\n")),
        Command(languages: ["es"], phrase: ["nueva", "línea"], action: .newlines("\n")),
        Command(languages: ["pt"], phrase: ["nova", "linha"], action: .newlines("\n")),
        Command(languages: ["es"], phrase: ["punto", "y", "coma"], action: .punctuation(";")),
        Command(languages: ["pt"], phrase: ["ponto", "e", "vírgula"], action: .punctuation(";")),
        Command(languages: ["es"], phrase: ["dos", "puntos"], action: .punctuation(":")),
        Command(languages: ["pt"], phrase: ["dois", "pontos"], action: .punctuation(":")),
        Command(languages: ["en"], phrase: ["no", "wait"], action: .removeLastWord),
        Command(languages: ["en"], phrase: ["i", "mean"], action: .removeLastWord),
        Command(languages: ["es"], phrase: ["mejor", "dicho"], action: .removeLastWord),
        Command(languages: ["es"], phrase: ["quise", "decir"], action: .removeLastWord),
        Command(languages: ["pt"], phrase: ["quer", "dizer"], action: .removeLastWord),
        Command(languages: ["pt"], phrase: ["melhor", "dizendo"], action: .removeLastWord),
        Command(languages: ["en"], phrase: ["scratch", "that"], action: .removeClause),
        Command(languages: ["es"], phrase: ["borra", "eso"], action: .removeClause),
        Command(languages: ["pt"], phrase: ["apaga", "isso"], action: .removeClause),
        Command(languages: ["en"], phrase: ["comma"], action: .punctuation(",")),
        Command(languages: ["en"], phrase: ["period"], action: .punctuation(".")),
        Command(languages: ["en"], phrase: ["colon"], action: .punctuation(":")),
        Command(languages: ["en"], phrase: ["semicolon"], action: .punctuation(";")),
        Command(languages: ["es"], phrase: ["coma"], action: .punctuation(",")),
        Command(languages: ["es"], phrase: ["punto"], action: .punctuation(".")),
        Command(languages: ["pt"], phrase: ["vírgula"], action: .punctuation(",")),
        Command(languages: ["pt"], phrase: ["ponto"], action: .punctuation("."))
    ]

    static func apply(_ transcript: String, language: String) -> String {
        let rawTokens = transcript.split(whereSeparator: { $0.isWhitespace }).map(String.init)
        guard !rawTokens.isEmpty else { return "" }
        let tokens = rawTokens.map(normalizeToken)
        let languageCode = normalizedLanguageCode(language)
        var output = ""
        var index = 0
        var literalMode = false
        var listIsNumbered = false
        var listNumber = 0
        var joinNext = false

        while index < rawTokens.count {
            let command = commands.first { candidate in
                guard languageCode == nil || candidate.languages.contains(languageCode!) else {
                    return false
                }
                let end = index + candidate.phrase.count
                guard end <= tokens.count else { return false }
                return Array(tokens[index..<end]) == candidate.phrase
            }

            if let command {
                if literalMode {
                    if case .literalOff = command.action {
                        literalMode = false
                        index += command.phrase.count
                        continue
                    }
                } else {
                    if case .literalOn = command.action {
                        literalMode = true
                        index += command.phrase.count
                        continue
                    }
                    apply(
                        command.action,
                        to: &output,
                        listIsNumbered: &listIsNumbered,
                        listNumber: &listNumber,
                        joinNext: &joinNext
                    )
                    index += command.phrase.count
                    continue
                }
            }

            if !literalMode, let languageCode,
               let entity = SpokenEntities.match(tokens, start: index, language: languageCode) {
                if !output.isEmpty && output.last?.isWhitespace == false && !joinNext {
                    output.append(" ")
                }
                output.append(entity.text)
                joinNext = false
                index += entity.consumed
                continue
            }

            if !output.isEmpty && output.last?.isWhitespace == false && !joinNext {
                output.append(" ")
            }
            output.append(rawTokens[index])
            joinNext = false
            index += 1
        }
        return output.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func normalizeToken(_ token: String) -> String {
        token.trimmingCharacters(in: CharacterSet.alphanumerics.inverted).lowercased()
    }

    private static func normalizedLanguageCode(_ language: String) -> String? {
        let normalized = language.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard normalized != "auto", normalized.count >= 2 else { return nil }
        let code = String(normalized.prefix(2))
        return ["en", "es", "pt"].contains(code) ? code : nil
    }

    private static func trimEnd(_ output: inout String) {
        while output.last?.isWhitespace == true {
            output.removeLast()
        }
    }

    private static func apply(
        _ action: Action,
        to output: inout String,
        listIsNumbered: inout Bool,
        listNumber: inout Int,
        joinNext: inout Bool
    ) {
        trimEnd(&output)
        switch action {
        case .punctuation(let mark):
            if !output.isEmpty && output.last != mark {
                output.append(mark)
            }
        case .newlines(let newlines):
            if !output.isEmpty {
                output.append(newlines)
            }
        case .removeLastWord:
            while output.last?.isWhitespace == false {
                output.removeLast()
            }
            trimEnd(&output)
        case .removeClause:
            if let boundary = output.lastIndex(where: { ".!?\n".contains($0) }) {
                output = String(output[...boundary])
            } else {
                output = ""
            }
            trimEnd(&output)
        case .open(let mark):
            if !output.isEmpty && output.last?.isWhitespace == false {
                output.append(" ")
            }
            output.append(mark)
            joinNext = true
        case .close(let mark):
            output.append(mark)
        case .join(let value):
            output.append(value)
            joinNext = true
        case .bullet:
            listIsNumbered = false
            listNumber = 0
            startListItem("- ", output: &output)
        case .numberedItem:
            if !listIsNumbered { listNumber = 0 }
            listIsNumbered = true
            listNumber += 1
            startListItem("\(listNumber). ", output: &output)
        case .nextItem:
            if listIsNumbered {
                listNumber += 1
                startListItem("\(listNumber). ", output: &output)
            } else {
                startListItem("- ", output: &output)
            }
        case .literalOn, .literalOff:
            break
        }
    }

    private static func startListItem(_ prefix: String, output: inout String) {
        trimEnd(&output)
        if !output.isEmpty { output.append("\n") }
        output.append(prefix)
    }
}
